lib/LANraragi/Controller/Api/Archive.pm

```diff
@@ -107,6 +114,173 @@ sub serve_file {
     $self->render_file( filepath => $file );
 }
 
+# Create a file archive along with any metadata.
+# adapted from Upload.pm
+sub create_archive {
+    my $self = shift;
+
+    my $logger = get_logger( "Archive API ", "lanraragi");
+    my $redis   = LANraragi::Model::Config->get_redis;
+
+    # receive uploaded file
+    my $upload              = $self->req->upload('file');
+    my $expected_checksum   = $self->req->param('file_checksum'); # optional
+
+    # require file
+    if ( ! defined $upload || !$upload ) {
+        return $self->render(
+            json => {
+                operation   => "upload",
+                success     => 0,
+                error       => "No file attached"
+            },
+            status => 400
+        );
+    }
+
+    # checksum verification stage.
+    if ( $expected_checksum ) {
+        my $file_content        = $upload->slurp;
+        my $actual_checksum     = sha1_hex($file_content);
+        if ( $expected_checksum ne $actual_checksum ) {
+            return $self->render(
+                json => {
+                    operation   => "upload",
+                    success     => 0,
+                    error       => "Checksum mismatch: expected $expected_checksum, got $actual_checksum."
+                },
+                status => 417
+            );
+        }
+    }
+
+    my $filename        = $upload->filename;
+    my $uploadMime      = $upload->headers->content_type;
+
+    # utf downgrade (see LANraragi::Utils::Minion)
+    $filename = encode('UTF-8', $filename);
+    unless (utf8::downgrade($filename, 1)) {
+        $logger->error("Bullshit! File path \"$filename\" could not be converted back to a byte sequence!");
+        return $self->render(
+            json => {
+                operation   => "upload",
+                success     => 0,
+                error       => "\"$filename\" could not be converted back to a byte sequence!"
+            },
+            status => 422
+        )
+    };
+
+    # lock resource
+    my $lock            = $redis->setnx( "upload:$filename", 1 );
+    if ( !$lock ) {
+        return $self->render(
+            json => {
+                operation   => "upload",
+                success     => 0,
+                error       => "Locked resource: $filename."
+            },
+            status => 423
+        );
+    }
+    $redis->expire( "upload:$filename", 10 );
+
+    # metadata extraction
+    my $catid           = $self->req->param('category_id');
+    my $tags            = $self->req->param('tags');
+    my $title           = $self->req->param('title');
+    my $summary         = $self->req->param('summary');
+
+    # return error if archive is not supported.
+    if ( !is_archive($filename) ) {
```

Comment:
You don't really need this bit if it's already done in `handle_incoming_file`, right? 

(I'm aware Controller->Upload does the same, but I think I'll eventually deprecate that code path in favor of just using the API in the webclient)

Reply:
`is_archive` ensures that the filename has a valid extension. I don't know how removing this will affect the tempdir logic, I'll need to test it out

Reply:
huh, I guess is_pdf wasn't written to handle invalid filenames at all and the issue just never popped up - good to know.